src/hidimstat/ensemble_clustered_inference.py

```diff
@@ -1,593 +1,446 @@
+    importances_ : ndarray, shape (n_features,) or (n_features, n_tasks)
+        Estimated coefficients at feature level.
+    pvalues_ : ndarray, shape (n_features,)
+        P-values for each feature.
+    one_minus_pvalues_ : ndarray, shape (n_features,)
+        One minus the p-values for numerical stability.
 
-    Notes
-    -----
-    The algorithm performs these steps for each bootstrap iteration:
-    1. Subsample the data using stratified sampling if groups are provided
-    2. Cluster features using Ward's hierarchical clustering
-    3. Transform data to reduced cluster space
-    4. Perform statistical inference using desparsified lasso
-    5. Aggregate results across all iterations
 
-    References
-    ----------
     .. footbibliography::
     """
-    memory = check_memory(memory=memory)
-    assert issubclass(
-        ward.__class__, FeatureAgglomeration
-    ), "ward need to an instance of sklearn.cluster.FeatureAgglomeration"
-    rng = check_random_state(random_state)
-
-    # Clustered inference algorithms
-    results = Parallel(n_jobs=n_jobs, verbose=verbose)(
-        delayed(clustered_inference)(
-            X_init,
-            y,
-            clone(ward),
-            scaler_sampling=scaler_sampling,
-            train_size=train_size,
-            groups=groups,
-            random_state=spawned_state,
-            n_jobs=1,
-            verbose=verbose,
-            memory=memory,
-            **kwargs,
+
+    def __init__(
+        self,
+        desparsified_lasso,
+        clustering,
+        n_bootstraps=25,
+        cluster_boostrap_size=0.3,
+        bootstrap_groups=None,
+        n_jobs=1,
+        random_state=None,
+        memory=None,
+        ensembling_method="quantiles",
+        gamma=0.5,
+        adaptive_aggregation=False,
+    ):
+        self.desparsified_lasso = desparsified_lasso
+        self.clustering = clustering
+        self.n_bootstraps = n_bootstraps
+        self.cluster_boostrap_size = cluster_boostrap_size
+        self.bootstrap_groups = bootstrap_groups
+        self.n_jobs = n_jobs
+        self.random_state = random_state
+        self.memory = memory
+        self.ensembling_method = ensembling_method
+        self.gamma = gamma
+        self.adaptive_aggregation = adaptive_aggregation
+
+        self.desparsified_lassos_ = None
+
+    def fit(self, X, y):
+        """
+        Fit multiple clustered inferences on random subsamples of the data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Input data matrix.
+        y : ndarray, shape (n_samples,) or (n_samples, n_tasks)
+            Target variable(s).
+
+        Returns
+        -------
+        self : EnCluDL
+            Fitted estimator.
+        """
+        rng = check_random_state(self.random_state)
+
+        def _fit_one(rng_spawned):
+            clu_dl = CluDL(
+                desparsified_lasso=deepcopy(self.desparsified_lasso),
```

Comment:
You should not use deepcopy but clone from sciklearn because deepcopy can't have some issue.